app/services/content_embed_service.rb

```diff
@@ -0,0 +1,83 @@
+class ContentEmbedService
+  include ApplicationHelper
+
+  SUPPORTED_DOCUMENT_TYPES = %w[contact].freeze
+  UUID_REGEX = /([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})/
+  EMBED_REGEX = /({{embed:(#{SUPPORTED_DOCUMENT_TYPES.join('|')}):#{UUID_REGEX}}})/
+
+  class EmbeddedEdition
+    attr_reader :embed_code, :document_type, :content_id
+
+    def initialize(embed_code:, document_type:, content_id:)
+      @embed_code = embed_code
+      @document_type = document_type
+      @content_id = content_id
+    end
+
+    def edition
+      @edition ||= Edition.with_document.find_by(
+        state: "published",
+        content_store: "live",
+        document_type: @document_type,
+        documents: { content_id: @content_id },
+      )
+    end
+  end
+
+  def initialize(body)
+    @body = body
+  end
+
+  def embedded_editions
+    @embedded_editions ||= @body.scan(EMBED_REGEX).map do |match|
+      EmbeddedEdition.new(
+        embed_code: match[0],
+        document_type: match[1],
+        content_id: match[2],
+      )
+    end
+  end
+
+  def fetch_links
+    embedded_editions.group_by(&:document_type).transform_values do |items|
```

Comment:
This assumes a 1-1 match of document_type -> link_type, but I wonder whether we should have a special link_type for embeds instead.

i.e. this

```mermaid
erDiagram
    Guide ||--o{ Contact : embed
```

Instead of this:

```mermaid
erDiagram
    Guide ||--o{ Contact : contact
```

Reply:
Oh good point!

Reply:
We'd need to add the new link type to all the schemas for it to be permitted. I think here: https://github.com/alphagov/publishing-api/blob/main/content_schemas/formats/shared/base_edition_links.jsonnet

Reply:
I am curious about implementation options that avoid using links. I'm very new to them in both theory and practice, so I don't have a strong sense of how good a fit they are yet. I'm digging through the docs now and trying to answer the question: are we in fact talking about introducing a new concept here? Link expansion and dependency resolution were what made me think of links for embedded content, but on reflection I'm not sure how central those mechanisms are to what a link actually is.

(For me this doesn't impact the value of this spike at all, by the way)

Reply:
For me, it's really appealing to use the existing concept of links. Partly because we'd dependency resolution "for free", and we'll want to be able to do that (i.e. when an embedded object changes, update all the pages which embed it). Partly because piggy backing on the documents / editions concept (which we have to do if we're using links) means we can benefit from solved problems like "how do I schedule updates of this thing?". But also generally, embedded content just feels like a specialisation of linked content to me.

@ryanb-gds recently updated [the link expansion docs](https://docs.publishing.service.gov.uk/repos/publishing-api/link-expansion.html), so those are 100% worth a read if you haven't already. There are certainly some problems with links as they currently work (link set links belonging to a content item, rather than an edition, edition links not being expanded recursively), but at least for the simple case of a direct embed, those won't be a problem.

All that said, it's quite possible I've overlooked some issue with links for this use case, and I'd be very interested in hearing about edge cases or things which we might not be able to achieve with this model.